proposals/extensions.md

```diff
@@ -399,54 +399,237 @@ These requirements need more refinement as implementation progresses, and may ne
+- Its name is unspeakable and determined based on the contents of the CLR-level signature. More details below.  
+- Its type parameters have normalized names (`T0`, `T1`, ...) and have no attributes.  
+- It is public and sealed.
+- It is marked with the `specialname` flag and an `[Extension]` attribute.  
+
+The content-based name of the extension grouping type is based on the CLR-level signature and includes the following:
+- The fully qualified CLR name of the type of the extension parameter.
+    - Type parameter names will be normalized to `T0`, `T1`, etc ... based on the order they appear in the type declaration. This means an extension block with a type parameter `Dictionary<TKey, TValue>` will result in the fully qualified name being `System.Collections.Generic.Dictionary<T0, T1>`.
+    - The fully qualified name will not include the containing assembly. It is common for types to be moved between assemblies and that should not break the xml doc references.
+- Constraints of type parameters will be included and sorted such that reordering them in source code does not change the name. Specifically:
+    - Type parameter constraints will be listed in declaration order. The constraints for the Nth type parameter will occur before the Nth+1 type parameter.
+    - Type constraints will be sorted by comparing the full names ordinally.
+    - Non-type constraints are ordered deterministically and are handled so as to avoid any ambiguity or collision with type constraints.
+- Since this does not include attributes, it intentionally ignores C#-isms like tuple names, nullability, etc ...
+
+Note: The name is guaranteed to remain stable across re-compilations, re-orderings and changes of C#-isms (ie. that don't affect the CLR-level signature).   
+
+#### Extension marker types
+
+The marker type re-declares the type parameters of its containing grouping type (an extension grouping type) to get full fidelity of the C# view of extension blocks.
+
+An extension marker type is emitted to metadata for each set of extension blocks in source with a the same C#-level signature.  
+- Its name is unspeakable and determined based on the contents of the C#-level signature of the extension block. More details below.  
+- It redeclares the type parameters for its containing grouping type to be those declared in source (including name and attributes).
+- It is private and static.
 - It is marked with the `specialname` flag.  
 
-Method/property declarations in an extension declaration in source are represented as members of the extension type in metadata.  
-The signatures of the original methods are maintained (including attributes), but their bodies are replaced with `throw null`.  
-Those should not be referenced in IL.  
+The content-based name of the extension marker type is based on the following:
+- Constraints of type parameters will be included and sorted such that reordering them in source code does not change the name.
```

Comment:
@jaredpar I'm thinking to relax this requirement. If a change to tuple names or nullability annotations affect the marker name, why not also let a re-ordering of constraints affect the marker name?